### PR TITLE
Sprint 4

### PR DESCRIPTION
**Pull request by andyocampo**
This branch adds a confetti celebration that happens each time a frog gets to the goal, to help indicate that they actually made it. This also adds a win/loss screen that lets the player know when the level has become unbeatable (losing more than half the frogs) or when they've succeeded.

Issues that have to be fixed: Cursor appears behind UI panels

### CODE REVIEW

File: Assets/Scripts/EndGoal.cs [hunk 1]
```diff
@@ -8,28 +8,34 @@ public class EndGoal : MonoBehaviour
 {
 
     private string frogsRemainingText;
-    private int frogGoal = 1;
-
-
-    void Update()
+    private int frogGoal = 0;
+    private int frogAmount = 0; //amount of frogs in level
+    public int FrogGoal
```
Comment (fatjosephina) [inline Assets/Scripts/EndGoal.cs:13]: It looks like FrogGoal is the amount of frogs which need to touch the goal in order to win and is not the same thing as the goal which the frogs are trying to reach (EndGoal). However, this is not super clear from the name. I am not sure what other name you could use, but I would recommend putting a summary whether you decide to change the name or not.

**Review by fatjosephina — APPROVED** [review]
All of the commits which are said to have been completed have been completed, and everything is running great in Unity. Given the number of open issues I do wonder if anything will have to be cut. I doubt it, but it may be something to think about. In terms of code, I only left one comment, but now that summaries and tool tips are required, those should be added to most of the main non-MonoBehaviour methods, properties, etc. Your hierarchy in Unity is organized very well. I also noticed a visual bug on the "ceiling" of the level which I hadn't before during play mode. There is a slight seam in between the tiles. I am not sure if this is fixable or not, but if not it is not that big of a deal. Here is a screenshot:
image
One last thing I wanted to mention is the scene title, Level 1. I don't think it matters too too much for the Alpha version, but between the sort of "cage" at the top and the mud, I think that this might be a little too involved for the first level of the game, unless maybe there is some sort of tutorial or other explanation to the player. Overall, great game as I said before, and the new commits are working well.

**Comment by fatjosephina** [comment]
One last comment to confirm: I'm merging the pull request, but I'll leave deleting the branch up to you.

**Comment by dantogno** [comment]
@fatjosephina  Thanks for looking at this closely!

File: Assets/Scripts/EndGoal.cs [hunk 2]
```diff
@@ -8,28 +8,34 @@ public class EndGoal : MonoBehaviour
 {
 
     private string frogsRemainingText;
-    private int frogGoal = 1;
-
-
-    void Update()
+    private int frogGoal = 0;
```
Comment (dantogno) [inline Assets/Scripts/EndGoal.cs:11]: Add summaries to the fields that may not be obvious.

File: Assets/Scripts/EndGoal.cs [hunk 3]
```diff
@@ -8,28 +8,34 @@ public class EndGoal : MonoBehaviour
...
     {
-        frogsRemainingText = $"Frogs: {frogGoal}/{FrogSpawner.amountOfFrogs}";
+            get { return frogGoal; }
     }
 
+    private ParticleSystem goalParticles; //confetti particle system
+
     //observer pattern that tells frog remaining UI how many frogs left
-    public static event Action<string> FrogReachesGoalTriggered; 
+    public static event Action<string> FrogReachesGoalTriggered;
```
Comment (dantogno) [inline Assets/Scripts/EndGoal.cs:21]: What is the actual event that happens here? The Frog has reached the goal? I think it's overkill to name it GoalReached and Triggered. I'd probably just go with `GoalReached` if that's accurate.

File: Assets/Scripts/EndGoal.cs [hunk 4]
```diff
@@ -8,28 +8,34 @@ public class EndGoal : MonoBehaviour
...
             Destroy(other.gameObject); //destroys frog
             frogGoal++;
 
-            if (frogGoal == FrogSpawner.amountOfFrogs + 1)
-            {
-                frogsRemainingText = "You Win!";
-            }
+            frogsRemainingText = $"Frogs: {frogGoal}/{frogAmount}";
 
             if (FrogReachesGoalTriggered != null)
```
Comment (dantogno) [inline Assets/Scripts/EndGoal.cs:40]: you can use the ?. syntax here to simplify and remove the if statement. https://thedotnetguide.com/null-conditional-operator-in-csharp/#

`FrogReachesGoalTriggered?.Invoke()`

File: Assets/Scripts/FrogSpawner.cs [hunk 5]
```diff
@@ -7,8 +7,8 @@ public class FrogSpawner : MonoBehaviour
     [SerializeField]
     GameObject frog;
 
-    public int setAmountOfFrogs;
-    public static int amountOfFrogs;
+    [SerializeField] int setAmountOfFrogs;
```
Comment (dantogno) [inline Assets/Scripts/FrogSpawner.cs:10]: add ToolTip attribute

File: Assets/Scripts/UI Scripts/WinLoseHandler.cs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    [SerializeField] GameObject hazards;
+    private int alive;
+    private int dead;
+
+
+    public static event Action<string> WinLoseTriggered;
+
+    void Update()
+    {
+        alive = goal.GetComponent<EndGoal>().FrogGoal;
```
Comment (dantogno) [inline Assets/Scripts/UI Scripts/WinLoseHandler.cs:19]: Don't use GetComponent in Update, it's really expensive. Do it in Start or Awake and cache the result (aka store it as a variable so you don't need to do it again)

File: Assets/Scripts/UI Scripts/WinLoseHandler.cs [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        dead = hazards.GetComponent<Hazard>().FrogsDead;
+
+        string WinText;
+
+        if (alive >= FrogSpawner.amountOfFrogs) //bad way of checking win
```
Comment (dantogno) [inline Assets/Scripts/UI Scripts/WinLoseHandler.cs:24]: Maybe this can all be handled by `EndGoal.cs`?

**Pull request merged**